src/Microsoft.AspNetCore.Server.Kestrel.Core/Internal/Http/Frame.FeatureCollection.cs

```diff
@@ -235,6 +235,13 @@ void IHttpResponseFeature.OnCompleted(Func<object, Task> callback, object state)
                 throw new InvalidOperationException(CoreStrings.CannotUpgradeNonUpgradableRequest);
             }
 
+            if (!ServiceContext.Resources.UpgradedConnections.TryLockOne())
```

Comment:
@DamianEdwards @davidfowl  You don't want upgraded connections applying to the total connection limit, right?